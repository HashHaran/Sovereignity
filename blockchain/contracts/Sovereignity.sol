// SPDX-License-Identifier: MIT
pragma solidity ^0.8.9;

// Import this file to use console.log
import "hardhat/console.sol";
import "./ISovereignity.sol";
import "../node_modules/@openzeppelin/contracts/utils/Context.sol";
import "../node_modules/@openzeppelin/contracts/security/ReentrancyGuard.sol";

contract Sovereignity is ISovereignity, Context, ReentrancyGuard {
    mapping(string => address) private _contentOwner;
    mapping(string => mapping(address => bool)) private _permissionedUsers;

    modifier _ownerOfContentOnly(string calldata contentId) {
        require(
            _contentOwner[contentId] != address(0),
            "Content not added yet"
        );
        require(
            _msgSender() == _contentOwner[contentId],
            "Not owner of content"
        );
        _;
    }

    function ownerOf(string calldata contentId)
        public
        view
        override
        returns (address)
    {
        return _contentOwner[contentId];
    }

    function addContent(string calldata contentId)
        external
        override
        nonReentrant
    {
        require(
            _contentOwner[contentId] == address(0),
            "Content already added and owned."
        );
        _contentOwner[contentId] = _msgSender();
        emit ContentCreation(
            contentId,
            contentId,
            _msgSender(),
            block.timestamp
        );
    }

    function deleteContent(
        string calldata contentId,
        address[] calldata permUsers
    ) external override _ownerOfContentOnly(contentId) nonReentrant {
        if (permUsers.length != 0) {
            _revokePermissionMulti(contentId, permUsers);
        }
        address currentOwner = _contentOwner[contentId];
        _contentOwner[contentId] = address(0);
        emit ContentDeletion(
            contentId,
            contentId,
            currentOwner,
            block.timestamp
        );
    }

    function isPermitted(string calldata contentId, address user)
        public
        view
        override
        returns (bool)
    {
        return
            _permissionedUsers[contentId][user] ||
            (user == _contentOwner[contentId]);
    }

    function awardPermission(string calldata contentId, address user)
        public
        override
        _ownerOfContentOnly(contentId)
        nonReentrant
    {
        require(
            !_permissionedUsers[contentId][user],
            "User already has permission to access content"
        );
        _permissionedUsers[contentId][user] = true;
        emit PermissionGranted(
            contentId,
            contentId,
            _msgSender(),
            user,
            block.timestamp
        );
    }

    function _revokePermission(string calldata contentId, address user)
        internal
    {
        require(
            _permissionedUsers[contentId][user],
            "User does not have permission to access content"
        );
        _permissionedUsers[contentId][user] = false;
        emit PermissionRevoked(
            contentId,
            contentId,
            _msgSender(),
            user,
            block.timestamp
        );
    }

    function revokePermission(string calldata contentId, address user)
        public
        override
        _ownerOfContentOnly(contentId)
        nonReentrant
    {
        _revokePermission(contentId, user);
    }

    function _revokePermissionMulti(
        string calldata contentId,
        address[] calldata users
    ) internal {
        for (uint256 i = 0; i < users.length; i++) {
            _revokePermission(contentId, users[i]);
        }
    }

    function transferOwnership(
        string calldata contentId,
        address newOwner,
        address[] calldata permUsers
    ) external override _ownerOfContentOnly(contentId) nonReentrant {
        require(
            newOwner != address(0),
            "Cannot transfer ownership to zero address"
        );
        require(
            newOwner != _contentOwner[contentId],
            "Cannot transfer ownership to self"
        );
        _revokePermissionMulti(contentId, permUsers);
        _contentOwner[contentId] = newOwner;
        emit OwnershipTransfered(
            contentId,
            contentId,
            _msgSender(),
            newOwner,
            block.timestamp
        );
    }
}
