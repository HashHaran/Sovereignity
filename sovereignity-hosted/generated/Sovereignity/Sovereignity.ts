// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.

import {
  ethereum,
  JSONValue,
  TypedMap,
  Entity,
  Bytes,
  Address,
  BigInt
} from "@graphprotocol/graph-ts";

export class ContentCreation extends ethereum.Event {
  get params(): ContentCreation__Params {
    return new ContentCreation__Params(this);
  }
}

export class ContentCreation__Params {
  _event: ContentCreation;

  constructor(event: ContentCreation) {
    this._event = event;
  }

  get contentId(): Bytes {
    return this._event.parameters[0].value.toBytes();
  }

  get contentIdString(): string {
    return this._event.parameters[1].value.toString();
  }

  get owner(): Address {
    return this._event.parameters[2].value.toAddress();
  }

  get timeStamp(): BigInt {
    return this._event.parameters[3].value.toBigInt();
  }
}

export class ContentDeletion extends ethereum.Event {
  get params(): ContentDeletion__Params {
    return new ContentDeletion__Params(this);
  }
}

export class ContentDeletion__Params {
  _event: ContentDeletion;

  constructor(event: ContentDeletion) {
    this._event = event;
  }

  get contentId(): Bytes {
    return this._event.parameters[0].value.toBytes();
  }

  get contentIdString(): string {
    return this._event.parameters[1].value.toString();
  }

  get owner(): Address {
    return this._event.parameters[2].value.toAddress();
  }

  get timeStamp(): BigInt {
    return this._event.parameters[3].value.toBigInt();
  }
}

export class OwnershipTransfered extends ethereum.Event {
  get params(): OwnershipTransfered__Params {
    return new OwnershipTransfered__Params(this);
  }
}

export class OwnershipTransfered__Params {
  _event: OwnershipTransfered;

  constructor(event: OwnershipTransfered) {
    this._event = event;
  }

  get contentId(): Bytes {
    return this._event.parameters[0].value.toBytes();
  }

  get contentIdString(): string {
    return this._event.parameters[1].value.toString();
  }

  get owner(): Address {
    return this._event.parameters[2].value.toAddress();
  }

  get newOwner(): Address {
    return this._event.parameters[3].value.toAddress();
  }

  get timeStamp(): BigInt {
    return this._event.parameters[4].value.toBigInt();
  }
}

export class PermissionGranted extends ethereum.Event {
  get params(): PermissionGranted__Params {
    return new PermissionGranted__Params(this);
  }
}

export class PermissionGranted__Params {
  _event: PermissionGranted;

  constructor(event: PermissionGranted) {
    this._event = event;
  }

  get contentId(): Bytes {
    return this._event.parameters[0].value.toBytes();
  }

  get contentIdString(): string {
    return this._event.parameters[1].value.toString();
  }

  get owner(): Address {
    return this._event.parameters[2].value.toAddress();
  }

  get user(): Address {
    return this._event.parameters[3].value.toAddress();
  }

  get timeStamp(): BigInt {
    return this._event.parameters[4].value.toBigInt();
  }
}

export class PermissionRevoked extends ethereum.Event {
  get params(): PermissionRevoked__Params {
    return new PermissionRevoked__Params(this);
  }
}

export class PermissionRevoked__Params {
  _event: PermissionRevoked;

  constructor(event: PermissionRevoked) {
    this._event = event;
  }

  get contentId(): Bytes {
    return this._event.parameters[0].value.toBytes();
  }

  get contentIdString(): string {
    return this._event.parameters[1].value.toString();
  }

  get owner(): Address {
    return this._event.parameters[2].value.toAddress();
  }

  get user(): Address {
    return this._event.parameters[3].value.toAddress();
  }

  get timeStamp(): BigInt {
    return this._event.parameters[4].value.toBigInt();
  }
}

export class Sovereignity extends ethereum.SmartContract {
  static bind(address: Address): Sovereignity {
    return new Sovereignity("Sovereignity", address);
  }

  isPermitted(contentId: string, user: Address): boolean {
    let result = super.call(
      "isPermitted",
      "isPermitted(string,address):(bool)",
      [ethereum.Value.fromString(contentId), ethereum.Value.fromAddress(user)]
    );

    return result[0].toBoolean();
  }

  try_isPermitted(
    contentId: string,
    user: Address
  ): ethereum.CallResult<boolean> {
    let result = super.tryCall(
      "isPermitted",
      "isPermitted(string,address):(bool)",
      [ethereum.Value.fromString(contentId), ethereum.Value.fromAddress(user)]
    );
    if (result.reverted) {
      return new ethereum.CallResult();
    }
    let value = result.value;
    return ethereum.CallResult.fromValue(value[0].toBoolean());
  }

  ownerOf(contentId: string): Address {
    let result = super.call("ownerOf", "ownerOf(string):(address)", [
      ethereum.Value.fromString(contentId)
    ]);

    return result[0].toAddress();
  }

  try_ownerOf(contentId: string): ethereum.CallResult<Address> {
    let result = super.tryCall("ownerOf", "ownerOf(string):(address)", [
      ethereum.Value.fromString(contentId)
    ]);
    if (result.reverted) {
      return new ethereum.CallResult();
    }
    let value = result.value;
    return ethereum.CallResult.fromValue(value[0].toAddress());
  }
}

export class AddContentCall extends ethereum.Call {
  get inputs(): AddContentCall__Inputs {
    return new AddContentCall__Inputs(this);
  }

  get outputs(): AddContentCall__Outputs {
    return new AddContentCall__Outputs(this);
  }
}

export class AddContentCall__Inputs {
  _call: AddContentCall;

  constructor(call: AddContentCall) {
    this._call = call;
  }

  get contentId(): string {
    return this._call.inputValues[0].value.toString();
  }
}

export class AddContentCall__Outputs {
  _call: AddContentCall;

  constructor(call: AddContentCall) {
    this._call = call;
  }
}

export class AwardPermissionCall extends ethereum.Call {
  get inputs(): AwardPermissionCall__Inputs {
    return new AwardPermissionCall__Inputs(this);
  }

  get outputs(): AwardPermissionCall__Outputs {
    return new AwardPermissionCall__Outputs(this);
  }
}

export class AwardPermissionCall__Inputs {
  _call: AwardPermissionCall;

  constructor(call: AwardPermissionCall) {
    this._call = call;
  }

  get contentId(): string {
    return this._call.inputValues[0].value.toString();
  }

  get user(): Address {
    return this._call.inputValues[1].value.toAddress();
  }
}

export class AwardPermissionCall__Outputs {
  _call: AwardPermissionCall;

  constructor(call: AwardPermissionCall) {
    this._call = call;
  }
}

export class DeleteContentCall extends ethereum.Call {
  get inputs(): DeleteContentCall__Inputs {
    return new DeleteContentCall__Inputs(this);
  }

  get outputs(): DeleteContentCall__Outputs {
    return new DeleteContentCall__Outputs(this);
  }
}

export class DeleteContentCall__Inputs {
  _call: DeleteContentCall;

  constructor(call: DeleteContentCall) {
    this._call = call;
  }

  get contentId(): string {
    return this._call.inputValues[0].value.toString();
  }

  get permUsers(): Array<Address> {
    return this._call.inputValues[1].value.toAddressArray();
  }
}

export class DeleteContentCall__Outputs {
  _call: DeleteContentCall;

  constructor(call: DeleteContentCall) {
    this._call = call;
  }
}

export class RevokePermissionCall extends ethereum.Call {
  get inputs(): RevokePermissionCall__Inputs {
    return new RevokePermissionCall__Inputs(this);
  }

  get outputs(): RevokePermissionCall__Outputs {
    return new RevokePermissionCall__Outputs(this);
  }
}

export class RevokePermissionCall__Inputs {
  _call: RevokePermissionCall;

  constructor(call: RevokePermissionCall) {
    this._call = call;
  }

  get contentId(): string {
    return this._call.inputValues[0].value.toString();
  }

  get user(): Address {
    return this._call.inputValues[1].value.toAddress();
  }
}

export class RevokePermissionCall__Outputs {
  _call: RevokePermissionCall;

  constructor(call: RevokePermissionCall) {
    this._call = call;
  }
}

export class TransferOwnershipCall extends ethereum.Call {
  get inputs(): TransferOwnershipCall__Inputs {
    return new TransferOwnershipCall__Inputs(this);
  }

  get outputs(): TransferOwnershipCall__Outputs {
    return new TransferOwnershipCall__Outputs(this);
  }
}

export class TransferOwnershipCall__Inputs {
  _call: TransferOwnershipCall;

  constructor(call: TransferOwnershipCall) {
    this._call = call;
  }

  get contentId(): string {
    return this._call.inputValues[0].value.toString();
  }

  get newOwner(): Address {
    return this._call.inputValues[1].value.toAddress();
  }

  get permUsers(): Array<Address> {
    return this._call.inputValues[2].value.toAddressArray();
  }
}

export class TransferOwnershipCall__Outputs {
  _call: TransferOwnershipCall;

  constructor(call: TransferOwnershipCall) {
    this._call = call;
  }
}
