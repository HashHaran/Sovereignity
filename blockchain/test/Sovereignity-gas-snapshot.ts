import snapshotGasCost from '@uniswap/snapshot-gas-cost'
import { ethers } from "hardhat";
import { Signer } from "@ethersproject/abstract-signer";
import { Wallet } from 'ethers'
import { Sovereignity } from "../typechain-types/Sovereignity";
var fs = require('fs');

describe('gas tests', () => {
    //Defining a minimalistic fixture for the Sovereignity, supporting all test cases
    let owner: Wallet, permittedUser: Wallet, otherUser: Wallet, nextOwner: Wallet;
    let sovereignity: Sovereignity;
    const contentId1 = "QmY7Yh4UquoXHLPFo2XbhXkhBvFoPwmQUSa92pxnxjQuPW";
    const contentId2 = "QmY7Yh4UquoXHLPFo2XbhXkhBvFoPwmQUSa92pxnaxHBKf";
    const contentId3 = "QmY7Yh4UquoXHLPFo2XbhXkhBvFoakdQUSa92pxnaxHBKf";
    async function sovereignityDeployedFixture() {
        var sovereignityAddress: {
            contractAddress: string;
        };
        sovereignityAddress = { contractAddress: "PLACEHOLDER" };
        var data = fs.readFileSync('contractAddress.json', 'utf8')
        sovereignityAddress = JSON.parse(data);
        // console.log(sovereignityAddress);

        const Sovereignity = await ethers.getContractFactory("Sovereignity");
        const sovereignity: Sovereignity = Sovereignity.attach(sovereignityAddress.contractAddress);

        const contentAdditionTx = await sovereignity.connect(owner as Signer).addContent(contentId1);
        await contentAdditionTx.wait();
        const awardPermissionTx = await sovereignity.connect(owner as Signer).awardPermission(contentId1, permittedUser.getAddress());
        await awardPermissionTx.wait();

        return sovereignity;
    }

    before('create fixture loader', async () => {
        [owner, permittedUser, otherUser, nextOwner] = await (ethers as any).getSigners();
        sovereignity = await sovereignityDeployedFixture();
    });
    
    it('gas snapshot for add content', async () => {
        await snapshotGasCost(sovereignity.connect(owner as Signer).addContent(contentId2));
    })
    
    it('gas snapshot for owner of', async () => {
        await snapshotGasCost(sovereignity.estimateGas.ownerOf(contentId1));
    })

    it('gas snapshot for award permission', async () => {
        await snapshotGasCost(sovereignity.connect(owner as Signer).awardPermission(contentId2, otherUser.getAddress()));
    })

    it('gas snapshot for revoke permission', async () => {
        await snapshotGasCost(sovereignity.connect(owner as Signer).revokePermission(contentId2, otherUser.getAddress()));
    })

    it('gas snapshot for is permitted', async () => {
        await snapshotGasCost(sovereignity.estimateGas.isPermitted(contentId1, permittedUser.getAddress()));
    })

    it('gas snapshot for transfer content', async () => {
        await snapshotGasCost(sovereignity.connect(owner as Signer).transferOwnership(contentId2, nextOwner.getAddress(), []));
    })

    it('gas snapshot for delete content', async () => {
        await snapshotGasCost(sovereignity.connect(nextOwner as Signer).deleteContent(contentId2, []));
    })
})