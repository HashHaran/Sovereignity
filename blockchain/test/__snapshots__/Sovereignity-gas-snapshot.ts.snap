// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gas tests gas snapshot for add content 1`] = `
Object {
  "calldataByteLength": 132,
  "gasUsed": 50816,
}
`;

exports[`gas tests gas snapshot for award permission 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 54504,
}
`;

exports[`gas tests gas snapshot for delete content 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 30150,
}
`;

exports[`gas tests gas snapshot for is permitted 1`] = `25524`;

exports[`gas tests gas snapshot for owner of 1`] = `25030`;

exports[`gas tests gas snapshot for revoke permission 1`] = `
Object {
  "calldataByteLength": 164,
  "gasUsed": 32616,
}
`;

exports[`gas tests gas snapshot for transfer content 1`] = `
Object {
  "calldataByteLength": 228,
  "gasUsed": 35877,
}
`;
